Fix TOS 3/4 VDI with memory protection

mikrosk opened the pull request:
This is my take on #22 . There were two separate issues with it:
- syscall context was no longer saved as expected: since the introduction of reentrant syscalls the PC in the `ctxt` struct wouldn't belong to the last caller (TOS VDI routine in our case) but to the user application calling `v_opnvwk` (i.e. the malloc() call would be made while [the in_kernel flag](https://github.com/freemint/freemint/blob/master/sys/arch/syscall.S#L699) is already true and not saving any context)
- xaaes.km is a kernel module and therefore all functions (incl. [its v_opnvwk](https://github.com/freemint/freemint/blob/master/xaaes/src.km/k_init.c#L726)) are always called with `in_kernel` flag set so no new context is built/saved either.

As you can see, I have solved it via an addition flag, `in_vdi`, which signalises whether the currently processed syscall is originating from a VDI call (trap 2 call to be precise but AES has its own dispatcher in the kernel so it's basically just VDI all along).

Additionally I have introduced a flag `tos_vdi` in kernel which ignores this workaround in case of NVDI/fVDI/NOVA VDI/EmuTOS VDI has been found.

Btw that `syscall.S` file could use some cleanup. We should decide what to do especially with the `NEW_SYSCALL` code, it makes everything super hard to read and follow.

th-otto commented:
Just a few comments:

- why did you remove the -s flag in the Makefile? IMHO a kernel with symbols is rather useless, since you won't be able to run it in a debugger.
- The check for NVDI/fVDI won't work. Both are started either from mint.cnf, or after mint, so their cookies are not yet set.

mikrosk commented:
> why did you remove the -s flag in the Makefile? IMHO a kernel with symbols is rather useless, since you won't be able to run it in a debugger.

Hatari debugger, anyone? ;) It has been a huge help when working on this issue it so it is a real world scenario (while you can strip the binary with `make strip`, you can "unstrip" it with this flag in place)

> The check for NVDI/fVDI won't work. Both are started either from mint.cnf, or after mint, so their cookies are not yet set.

Good catch! I wanted to optimise it a little (so the check wouldn't be done for every M(x)alloc call). I'll change it.

> Previously, without saving the context for the Malloc, the pc was checked for being in ROM. This was the PC the v_opnvwk call, and therefor exactly what was needed. I quite don't understand why the patch should fix the problem (although is a bit cleaner that checking for absolute ROM addresses).

It's because of that re-entrancy. Before this feature had been implemented by Ozk, all calls were quite straightforward -- an application makes a syscall, we store PC of the caller. If it's chained (`v_opnvwk` -> `Malloc`), the caller is ROM because this is where `v_opnvwk` would be implemented. However now the kernel stores the application context by the first call (`v_opnvwk`) and by the next call (`Malloc`) kernel directly jumps to its re-entrant handler, without storing anything. So `ctxt` still contains the application PC and not ROM PC. And for `xaaes.km` it is even more complicated, see above.

> I agree with the NEW_SYSCALL define. But that would probably mean we have to remove it, since that code was never tested.

I guess it was tested by Konrad back in the 2000s but its introduction by commit 54a7481f doesn't tell much about its purpose.

th-otto commented:
> So ctxt still contains the application PC and not ROM PC.

Yes, but that is intended in this case. You want to catch the v_opnvwk case, not the Malloc case.

mikrosk commented:
> Yes, but that is intended in this case. You want to catch the v_opnvwk case, not the Malloc case.

Not sure what you mean? `v_opnvwk` is a normal application call, so naturally its origin is from application memory. On the contrary, we want to catch the `Malloc` case which is called from ROM's `v_opnvwk` implementation.

th-otto commented:
About the NEW_SYSCALL: the handler copies the trap parameters to the kernel stack. currently it does that blindly by copying 32 bytes, with NEW_SYSCALL only the actual needed parameters are copied. That may prevent problems with MP when SP+32 is maybe outside of the application stack. But i don't think it can be cleanly done that way, there are calls like VSetscreen/Rwabs that take extra parameters only in certain cases, and there may also be resident programs that install new functions which the kernel does not even know about.

mikrosk commented:
Ah, right, linking #116 for reference.

mikrosk commented:
I have implemented a few optimisations (like completely disabling this mechanism on plain 68K) and also simplified the condition to avoid unnecessary `get_cookie` calls. I was tempted to remove the check completely (i.e. rely on VDI to supply a non-zero `mode`) but then I realised that VDI could make another M(x)alloc call which would be completely fine to have a private memory and we would force it to be 'super accessible' for no reason.
